### ISSUE TITLE
[Feature] Introduce Merge Engines for Primary-Key Table

### ISSUE THREAD

**Issue by wuchong** [alibaba/fluss#212]
### Search before asking

- [X] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.

### Motivation

We propose the introduction of merge engines for primary-key tables, enhancing data management and query capabilities. This feature will encompass the following merge engines, each with distinct functionalities:

- Version-Based Merge: This merge engine will allow data consolidation based on version numbers. It is very similar to ClickHouse [ReplacingMergeTree](https://clickhouse.com/docs/en/guides/replacing-merge-tree) that keeps the row with the highest version number. This is also very similar to HBase [checkAndPut](https://hbase.apache.org/1.2/apidocs/org/apache/hadoop/hbase/client/Table.html#checkAndPut(byte[],%20byte[],%20byte[],%20byte[],%20org.apache.hadoop.hbase.client.Put)) that updates records only when the condition is true. In this way, the out-of-order data can be guaranteed to be ultimately consistent with the upstream. 

- First-Row Based Merge: Designed to retain the first occurrence of a row with the same primary key, this engine boosts performance by minimizing writes to the kv store and generates only insert-only changelogs (some downstream jobs only accept insert-only changelogs). 

- Aggregate Table Merge: This engine will facilitate the merging of data by aggregating records, which is particularly useful for summarizing information and analytics purposes. It will support operations like `sum`, `max`, `min`, `count`, `avg`, among others, providing robust analytic capabilities.

### Solution

Introduce table properties (configured via Flink DDL with options):
```properties
'table.merge-engine' = 'first_row|high_version|aggregate'
'table.merge-engine.version.column' = '<version_column>'
'table.merge-engine.aggregate.<agg_column>' =  'sum | max | min | count | avg | first_value | last_value' 
```

The `merge-engine` properties are table storage properties, and therefore can't be changed after table creation. 

### Version-Based Merge

```sql
CREATE TABLE fluss_table (
  id BIGINT,
  version BIGINT,
  data STRING,
  PRIMARY KEY (id) NOT ENFORCED
) WITH (
  'connector' = 'fluss',
  'bootstrap.address' = '...',
  'table.merge-engine' = 'high_version',
  'table.merge-engine.version.column' = 'version'  -- updates when new version >= old version  
);
```

### First-Row-Based Merge

```sql
CREATE TABLE fluss_table (
  id BIGINT,
  version BIGINT,
  data STRING,
  PRIMARY KEY (id) NOT ENFORCED
) WITH (
  'connector' = 'fluss',
  'bootstrap.address' = '...',
  'table.merge-engine' = 'first_row' -- only the first row of the primary key will be retained.
);
```

### Aggregate-Based Merge

```sql
CREATE TABLE fluss_table (
    id BIGINT,
    price DOUBLE,
    sales BIGINT,
    PRIMARY KEY (id) NOT ENFORCED
) WITH (
    'table.merge-engine' = 'aggregate',
    'table.merge-engine.aggregate.price' = 'max',
    'table.merge-engine.aggregate.sales' = 'sum'
);
```

### Anything else?

Subtasks: 

- [ ] Introduce first-row merge engine for primary key table
- [ ] Introduce high-version merge engine for primary key table
- [ ] Introduce aggregate merge engine for primary key table

### Willingness to contribute

- [ ] I'm willing to submit a PR!

**Comment by sunxiaojian** [comment]
HI  ,  @wuchong Have you started working on the Aggregate Merge Engine?  If not, I would like to continue working on it.

**Comment by wuchong** [comment]
Hi @sunxiaojian,

To ensure the Aggregate Merge Engine is idempotent, we need to introduce a distributed transaction mechanism . Without this, the aggregate results may become inconsistent if the Flink job restarts and duplicate data is written into Fluss.

Thus, implementing support for the Aggregate Merge Engine is a significant work, and we are currently in the design phase for the transaction mechanism.

**Comment by sunxiaojian** [comment]
> Hi [@sunxiaojian](https://github.com/sunxiaojian),
> 
> To ensure the Aggregate Merge Engine is idempotent, we need to introduce a distributed transaction mechanism . Without this, the aggregate results may become inconsistent if the Flink job restarts and duplicate data is written into Fluss.
> 
> Thus, implementing support for the Aggregate Merge Engine is a significant work, and we are currently in the design phase for the transaction mechanism.

@wuchong Yes, we need to introduce a distributed transaction mechanism to ensure idempotency,   just like in Kafka. https://www.confluent.io/blog/transactions-apache-kafka/  ,  will the distributed transaction also be sent to the community for joint development?

**Comment by wuchong** [comment]
@sunxiaojian yes, but maybe a different implementation rather than Kafka.

**Comment by sunxiaojian** [comment]
> [@sunxiaojian](https://github.com/sunxiaojian) yes, but maybe a different implementation rather than Kafka.

ok, thanks for your reply.

**Comment by simonsssu** [comment]
hi, @wuchong is txn design doc ready or still in discussion.
